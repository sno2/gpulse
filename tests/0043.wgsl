var<private> next_item: i32 = 0;

fn advance_item() -> i32 {
   next_item += 1;   // Adds 1 to next_item.
   return next_item - 1;
}

fn bump_item() {
  var data: array<f32,10>;
  next_item = 0;
  // Adds 5.0 to data[0], calling advance_item() only once.
  data[advance_item()] += 5.0;
  // next_item will be 1 here.
}

fn precedence_example() {
  var value = 1;
  // The right-hand side of a compound assignment is its own expression.
  value *= 2 + 3; // Same as value = value * (2 + 3);
  // 'value' now holds 5.
}
