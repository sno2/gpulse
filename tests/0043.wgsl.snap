ast.Node{ .var_decl = ast.VarDecl{ .name = ast.Span{ .start = 13, .end = 22 }, .access_mode = ast.AccessMode.read, .addr_space = ast.AddrSpace.private, .typ = ast.Node{ .identifier = ast.Span{ .start = 24, .end = 27 } }, .value = ast.Node{ .number_literal = ast.Span{ .start = 30, .end = 31 } } } } ast.Node{ .fn_decl = ast.FnDecl{ .name = ast.Span{ .start = 37, .end = 49 }, .params = {  }, .ret = ast.Node{ .identifier = ast.Span{ .start = 55, .end = 58 } }, .scope = { ast.Node{ .add_assign = ast.BinaryOp{ .lhs = ast.Node{ .identifier = ast.Span{ .start = 64, .end = 73 } }, .rhs = ast.Node{ .number_literal = ast.Span{ .start = 77, .end = 78 } } } }, ast.Node{ .ret = ast.Ret{ .value = ast.Node{ .sub = ast.BinaryOp{ .lhs = ast.Node{ .identifier = ast.Span{ .start = 116, .end = 125 } }, .rhs = ast.Node{ .number_literal = ast.Span{ .start = 128, .end = 129 } } } } } } } } } ast.Node{ .fn_decl = ast.FnDecl{ .name = ast.Span{ .start = 137, .end = 146 }, .params = {  }, .ret = null, .scope = { ast.Node{ .var_decl = ast.VarDecl{ .name = ast.Span{ .start = 157, .end = 161 }, .access_mode = ast.AccessMode.read, .addr_space = null, .typ = ast.Node{ .template = ast.Template{ .name = ast.Span{ .start = 163, .end = 168 }, .args = { ast.Node{ .identifier = ast.Span{ .start = 169, .end = 172 } }, ast.Node{ .number_literal = ast.Span{ .start = 173, .end = 175 } } } } }, .value = null } }, ast.Node{ .assign = ast.BinaryOp{ .lhs = ast.Node{ .identifier = ast.Span{ .start = 180, .end = 189 } }, .rhs = ast.Node{ .number_literal = ast.Span{ .start = 192, .end = 193 } } } }, ast.Node{ .add_assign = ast.BinaryOp{ .lhs = ast.Node{ .index = ast.BinaryOp{ .lhs = ast.Node{ .identifier = ast.Span{ .start = 257, .end = 261 } }, .rhs = ast.Node{ .call = ast.Call{ .callee = ast.Node{ .identifier = ast.Span{ .start = 262, .end = 274 } }, .args = {  } } } } }, .rhs = ast.Node{ .number_literal = ast.Span{ .start = 281, .end = 284 } } } } } } } ast.Node{ .fn_decl = ast.FnDecl{ .name = ast.Span{ .start = 323, .end = 341 }, .params = {  }, .ret = null, .scope = { ast.Node{ .var_decl = ast.VarDecl{ .name = ast.Span{ .start = 352, .end = 357 }, .access_mode = ast.AccessMode.read, .addr_space = null, .typ = null, .value = ast.Node{ .number_literal = ast.Span{ .start = 360, .end = 361 } } } }, ast.Node{ .mul_assign = ast.BinaryOp{ .lhs = ast.Node{ .identifier = ast.Span{ .start = 438, .end = 443 } }, .rhs = ast.Node{ .add = ast.BinaryOp{ .lhs = ast.Node{ .number_literal = ast.Span{ .start = 447, .end = 448 } }, .rhs = ast.Node{ .number_literal = ast.Span{ .start = 451, .end = 452 } } } } } } } } } 